StorageWriter fails post-recovery with partially written appends

**Problem description**
We have observed the following exception in the segment store while running some tests:

```
2017-12-07 18:35:51,305 80300529 [segment-store-3] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[9376].Error.
java.lang.IllegalArgumentException: startOffset must refer to an offset beyond the Segment's StorageLength offset.
    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
    at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.readDirect(StreamSegmentReadIndex.java:581)
    at io.pravega.segmentstore.server.reading.ContainerReadIndex.readDirect(ContainerReadIndex.java:176)
    at io.pravega.segmentstore.server.writer.StorageWriterFactory$StorageWriterDataSource.getAppendData(StorageWriterFactory.java:123)
    at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:649)
    at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:598)
    at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushExcess$14(SegmentAggregator.java:579)
    at io.pravega.common.concurrent.FutureHelpers$Loop.call(FutureHelpers.java:642)
    at io.pravega.common.concurrent.FutureHelpers$Loop.call(FutureHelpers.java:611)
    at io.pravega.common.concurrent.FutureHelpers.runOrFail(FutureHelpers.java:524)
    at io.pravega.common.concurrent.FutureHelpers$Loop.run(FutureHelpers.java:655)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

This error has been found against our 0.1 release, so it is possible that it is not an issue any longer, but I couldn't find a related issue. Unfortunately, we don't have for now any more information that this stack trace for now, although it seems to be related to an issue with BookKeeper.

**Problem location**
Segment store.

**Suggestions for an improvement**
Determine whether it is an issue or not and fix in the case it is.

This error is observed in the perf test runs too. The scneario there is that BK becomes unavailable for some time during which following OOM is observed. The segmentstore restarts and then above mentioned exception is observed. It is also observed that the segmentstore recovers from this error.
```
2017-12-12 03:38:00,684 39987486 [epollEventLoopGroup-3-2] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection: 
io.netty.handler.codec.DecoderException: java.lang.OutOfMemoryError: GC overhead limit exceeded
    at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:459)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
    at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
    at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
    at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
    at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
    at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:797)
    at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:404)
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:304)
    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
```

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
@@ -386,6 +386,13 @@ class SegmentAggregator implements OperationProcessor, AutoCloseable {
      */
     private void aggregateAppendOperation(StorageOperation operation, AggregatedAppendOperation aggregatedAppend, int maxLength) {
         long remainingLength = operation.getLength();
+        if (operation.getStreamSegmentOffset() < aggregatedAppend.getLastStreamSegmentOffset()) {
+            // The given operation begins before the AggregatedAppendOperation. This is likely due to it having been
+            // partially written to Storage prior to some recovery event. We must make sure we only include the part that
+            // has not yet been written.
+            remainingLength -= aggregatedAppend.getLastStreamSegmentOffset() - operation.getStreamSegmentOffset();
+        }
+
         while (remainingLength > 0) {
             // All append lengths are integers, so it's safe to cast here.
             int lengthToAdd = (int) Math.min(maxLength - aggregatedAppend.getLength(), remainingLength);
@@ -428,8 +435,10 @@ class SegmentAggregator implements OperationProcessor, AutoCloseable {
         }
 
         if (aggregatedAppend == null) {
-            // No operations or last operation not an AggregatedAppend - create a new one.
-            aggregatedAppend = new AggregatedAppendOperation(this.metadata.getId(), operationOffset, operationSequenceNumber);
+            // No operations or last operation not an AggregatedAppend - create a new one, while making sure the first
+            // offset is not below the current StorageLength (otherwise we risk re-writing data that's already in Storage).
+            long offset = Math.max(operationOffset, this.metadata.getStorageLength());
+            aggregatedAppend = new AggregatedAppendOperation(this.metadata.getId(), offset, operationSequenceNumber);
             this.operations.add(aggregatedAppend);
         }
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java
@@ -390,7 +390,9 @@ class SegmentAggregator implements OperationProcessor, AutoCloseable {
             // The given operation begins before the AggregatedAppendOperation. This is likely due to it having been
             // partially written to Storage prior to some recovery event. We must make sure we only include the part that
             // has not yet been written.
-            remainingLength -= aggregatedAppend.getLastStreamSegmentOffset() - operation.getStreamSegmentOffset();
+            long delta = aggregatedAppend.getLastStreamSegmentOffset() - operation.getStreamSegmentOffset();
+            remainingLength -= delta;
+            log.debug("Skipping {} bytes from the beginning of '{}' since it has already been partially written to Storage.", this.traceObjectId, delta, operation);
         }
 
         while (remainingLength > 0) {

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/SegmentAggregatorTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/SegmentAggregatorTests.java
@@ -888,7 +888,7 @@ public class SegmentAggregatorTests extends ThreadPooledTestSuite {
         }
 
         // Verify that in the end, the contents of the parents is as expected.
-        verifyParentSegmentData(parentData, context);
+        verifySegmentData(parentData.toByteArray(), context);
 
         // Verify calls to completeMerge.
         val expectedMergeOpSources = Arrays.stream(context.transactionAggregators).map(a -> a.getMetadata().getId()).collect(Collectors.toSet());
@@ -1655,9 +1655,55 @@ public class SegmentAggregatorTests extends ThreadPooledTestSuite {
         flushAllSegments(context);
 
         // Verify that in the end, the contents of the parents is as expected.
-        verifyParentSegmentData(parentData, context);
+        verifySegmentData(parentData.toByteArray(), context);
         AssertExtensions.assertListEquals("Unexpected callback calls to completeMerge for already processed operations.",
                 expectedMergeOpAck, actualMergeOpAck, Map.Entry::equals);
+    }
+
+    /**
+     * Tests a scenario where data that is about to be added already partially exists in Storage. This would most likely
+     * happen in a recovery situation, where we committed a part of an append operation before failing over.
+     */
+    @Test
+    public void testRecoveryPartialWrite() throws Exception {
+        final int writeLength = 1024;
+        final int partialWriteLength = writeLength / 2;
+
+        @Cleanup
+        TestContext context = new TestContext(DEFAULT_CONFIG);
+        context.storage.create(context.segmentAggregator.getMetadata().getName(), TIMEOUT).join();
+
+        // Store written data by segment - so we can check it later.
+        ArrayList<StorageOperation> operations = new ArrayList<>();
+
+        byte[] writtenData = new byte[writeLength];
+        val rnd = new Random(0);
+        rnd.nextBytes(writtenData);
+        StorageOperation appendOp = generateAppendAndUpdateMetadata(context.segmentAggregator.getMetadata().getId(), writtenData, context);
+        operations.add(appendOp);
+        operations.add(generateSealAndUpdateMetadata(context.segmentAggregator.getMetadata().getId(), context));
+
+        // Write half of the data to Storage.
+        context.storage.write(
+                writeHandle(context.segmentAggregator.getMetadata().getName()),
+                0,
+                new ByteArrayInputStream(writtenData),
+                partialWriteLength,
+                TIMEOUT).join();
+
+        // Initialize the SegmentAggregator. This should pick up the half-written operation.
+        context.segmentAggregator.initialize(TIMEOUT, executorService()).join();
+        Assert.assertEquals("", partialWriteLength, context.segmentAggregator.getMetadata().getStorageLength());
+
+        // Add all operations we had so far.
+        for (StorageOperation o : operations) {
+            context.segmentAggregator.add(o);
+        }
+
+        flushAllSegments(context);
+
+        // Verify that in the end, the contents of the parents is as expected.
+        verifySegmentData(writtenData, context);
     }
 
     //endregion
@@ -1804,8 +1850,7 @@ public class SegmentAggregatorTests extends ThreadPooledTestSuite {
         }
     }
 
-    private void verifyParentSegmentData(ByteArrayOutputStream parentData, TestContext context) {
-        byte[] expectedData = parentData.toByteArray();
+    private void verifySegmentData(byte[] expectedData, TestContext context) {
         byte[] actualData = new byte[expectedData.length];
         long storageLength = context.storage.getStreamSegmentInfo(context.segmentAggregator.getMetadata().getName(), TIMEOUT).join().getLength();
         Assert.assertEquals("Unexpected number of bytes flushed/merged to Storage.", expectedData.length, storageLength);

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/TestWriterDataSource.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/TestWriterDataSource.java
@@ -15,6 +15,7 @@ import io.pravega.common.Exceptions;
 import io.pravega.common.concurrent.Futures;
 import io.pravega.common.function.Callbacks;
 import io.pravega.common.util.SequencedItemList;
+import io.pravega.segmentstore.server.SegmentMetadata;
 import io.pravega.segmentstore.server.UpdateableContainerMetadata;
 import io.pravega.segmentstore.server.UpdateableSegmentMetadata;
 import io.pravega.segmentstore.server.logs.operations.MetadataCheckpointOperation;
@@ -273,6 +274,16 @@ class TestWriterDataSource implements WriterDataSource, AutoCloseable {
         }
 
         synchronized (this.lock) {
+            // Perform the same validation checks as the ReadIndex would do.
+            SegmentMetadata sm = this.metadata.getStreamSegmentMetadata(streamSegmentId);
+            Preconditions.checkArgument(length >= 0, "length must be a non-negative number");
+            Preconditions.checkArgument(startOffset >= sm.getStorageLength(),
+                    "startOffset must be larger than refer to an offset beyond the Segment's StorageLength offset.");
+            Preconditions.checkArgument(startOffset + length <= sm.getLength(),
+                    "startOffset+length must be less than the length of the Segment.");
+            Preconditions.checkArgument(startOffset >= Math.min(sm.getStartOffset(), sm.getStorageLength()),
+                    "startOffset is before the Segment's StartOffset.");
+
             ad = this.appendData.getOrDefault(streamSegmentId, null);
         }
 

